sfn-query-large-datasets-sam/template.yaml

```diff
@@ -0,0 +1,155 @@
+AWSTemplateFormatVersion: '2010-09-09'
+Transform: AWS::Serverless-2016-10-31
+Description: Query large datasets (Amazon Athena, Amazon S3, AWS Glue, Amazon SNS)
+
+Resources:
+  S3QueryLargeDatasetsBucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !Sub 'stepfunctions-query-large-datasets'
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+          - ServerSideEncryptionByDefault:
+              SSEAlgorithm: AES256
+
+  AthenaUpdatesTopic:
+    Type: AWS::SNS::Topic
+    Properties:
+      TopicName: 'athena-updates-topic'
+      KmsMasterKeyId: 'alias/aws/sns'
+
+  GlueCrawlerRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service: 'glue.amazonaws.com'
+            Action: 'sts:AssumeRole'
+      ManagedPolicyArns:
+        - 'arn:aws:iam::aws:policy/service-role/AWSGlueServiceRole'
+      RoleName: !Sub 'glue-crawler-role'
+
+  QueryLargeDatasetsDb:
+    Type: AWS::Glue::Database
+    Properties:
+      CatalogId: !Ref "AWS::AccountId"
+      DatabaseInput:
+        Name: !Sub 'query_large_datasets_db'
+        CreateTableDefaultPermissions:
+          - Permissions:
+              - 'ALL'
+            Principal:
+              DataLakePrincipalIdentifier: 'IAM_ALLOWED_PRINCIPALS'
+        Description: 'Database for large datasets'
+
+  StartAthenaCrawler:
+    Type: AWS::Glue::Crawler
+    Properties:
+      Role: !GetAtt GlueCrawlerRole.Arn
+      Targets:
+        S3Targets:
+          - Path: 's3://aws-glue-datasets-us-east-1/examples/githubarchive/month/data/'
+      Name: !Sub 'query_large_datasets_crawler'
+      DatabaseName: !Ref QueryLargeDatasetsDb
+      Configuration: !Sub '{"Version":1.0,"CrawlerOutput":{"Partitions":{"AddOrUpdateBehavior":"InheritFromTable"}},"Grouping":{"TableGroupingPolicy":"CombineCompatibleSchemas"}}'
+
+  AthenaWorkGroup:
+    Type: AWS::Athena::WorkGroup
+    Properties:
+      Name: !Sub 'log'
+      State: 'ENABLED'
+      RecursiveDeleteOption: 'true'
+      WorkGroupConfiguration:
+        ResultConfiguration:
+          OutputLocation: !Sub 's3://${S3QueryLargeDatasetsBucket}/result/'
+
+  StateMachineRole:
```

Comment:
In the CDK project, we are also giving these permissions to this role:

```python
iam.PolicyStatement(
    actions=[
        "s3:ListBucket",
        "s3:GetObject"
    ],
    resources=[
        f"arn:aws:s3:::aws-glue-datasets-us-east-1",
        f"arn:aws:s3:::aws-glue-datasets-us-east-1/*",
    ],
    effect=iam.Effect.ALLOW
),
```

Does this mean they were not really needed or we are missing them here?

Reply:
SAM version is 1:1 like the terraform from permissions point of view. for CDK I've hardened the security to also advertise the use of CDK-NAG.